### PR TITLE
Support terminal renaming inline in the tabs view

### LINKED ISSUE #122023: Allow renaming terminal tabs inside the tab view
Terminal now:

Screen Shot 2021-04-23 at 7 02 49 AM

What we want:

Screen Shot 2021-04-23 at 7 03 19 AM

### PR DESCRIPTION
**Pull request by meganrogge**
This PR fixes #122023

### CODE REVIEW

**Comment by meganrogge** [comment]
To do:

- where'd the inline actions go?
- renaming of single tab

File: src/vs/workbench/contrib/terminal/browser/terminalActions.ts [hunk 1]
```diff
@@ -1830,3 +1851,14 @@ function focusNext(accessor: ServicesAccessor): void {
 	const listService = accessor.get(IListService);
 	listService.lastFocusedList?.focusNext();
 }
+
+export function validateTerminalName(name: string): { content: string, severity: Severity } | null {
+	if (!name || name.length === 0 || /^\s+$/.test(name)) {
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalActions.ts:1856]:
```suggestion
	if (!name || name.trim().length === 0) {
```

File: src/vs/workbench/contrib/terminal/browser/terminalInstance.ts [hunk 2]
```diff
@@ -1789,7 +1789,10 @@ export class TerminalInstance extends Disposable implements ITerminalInstance {
 		return this._linkManager.registerExternalLinkProvider(this, provider);
 	}
 
-	async rename() {
+	async rename(title?: string) {
+		if (title) {
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalInstance.ts:1793]: Can flip this around:

```
if (!title) {
  title = await ...
}
if (title) {
  set...
}
```

File: src/vs/workbench/contrib/terminal/browser/terminalService.ts [hunk 3]
```diff
@@ -316,6 +318,27 @@ export class TerminalService implements ITerminalService {
...
+		if (!data) {
+			this._editable = undefined;
+		} else {
+			this._editable = { stat, data };
+		}
+		const pane = this._viewsService.getActiveViewWithId<TerminalViewPane>(TERMINAL_VIEW_ID);
+		const isEditing = this._isEditable(stat);
+		if (pane?.terminalTabbedView) {
+			pane.terminalTabbedView.setEditable(isEditing);
+		}
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalService.ts:331]: Can we combine using `?.`?

File: src/vs/workbench/contrib/terminal/browser/terminalService.ts [hunk 4]
```diff
@@ -69,6 +69,8 @@ export class TerminalService implements ITerminalService {
 	private _localTerminalsInitPromise: Promise<void> | undefined;
 	private _connectionState: TerminalConnectionState;
 
+	private _editable: { stat: ITerminalInstance, data: IEditableData } | undefined;
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalService.ts:72]: stat -> instance?

File: src/vs/workbench/contrib/terminal/browser/terminalTabsList.ts [hunk 5]
```diff
@@ -231,95 +239,184 @@ class TerminalTabsRenderer implements IListRenderer<ITerminalInstance, ITerminal
 	}
 
 	renderElement(instance: ITerminalInstance, index: number, template: ITerminalTabEntryTemplate): void {
-		const group = this._terminalService.getGroupForInstance(instance);
-		if (!group) {
-			throw new Error(`Could not find group for instance "${instance.instanceId}"`);
-		}
+		const editableData = this._terminalService.getEditableData(instance);
+
+		if (!editableData) {
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/terminalTabsList.ts:244]: Let's invert and return early

**Review by Tyriar — APPROVED** [review]

**Pull request merged**